diff -ru brlcad/src/libbn/mat.c brlcad-7.22.0/src/libbn/mat.c
--- brlcad/src/libbn/mat.c	2013-07-17 16:13:36.522008144 +0100
+++ brlcad-7.22.0/src/libbn/mat.c	2012-06-25 04:37:12.000000000 +0100
@@ -293,30 +293,9 @@
     int z[4];		/* Temporary */
     fastf_t b[4];	/* Temporary */
     fastf_t c[4];	/* Temporary */
-    mat_t temp_mat = MAT_INIT_ZERO;/*Temporary matrix to check for orthogonality.*/
-
 
     MAT_COPY(output, input);	/* Duplicate */
 
-
-    /**
-     *           ORTHOGONAL MATRICES
-     *@brief
-     *checks for orthogonality where( M * Transpose(M) = I)
-     *if so inverse becomes Inv(M) = Transpose(M)
-     *
-     */
-     
-    MAT_TRANSPOSE(temp_mat,output);/*sets temp_mat as transpose*/
-    bn_mat_mul2(output,temp_mat);
-
-    if(bn_mat_is_identity(temp_mat)) {
-	MAT_COPY(output,temp_mat);
-	return 1;
-    }
-
-
-    
     /* Initialization */
     for (j = 0; j < 4; j++)
 	z[j] = j;
